marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java

```diff
@@ -114,4 +116,52 @@ private String convertRowToJSONString(InternalRow record) {
         return jsonObjectWriter.toString();
     }
 
+    protected JsonNode determineIngestionEndpointParams(Map<String, String> params) {
+        ObjectMapper objectMapper = new ObjectMapper();
+
+        String ingestEndpointParams = params.containsKey("ingestendpointparams")?
+            params.get("ingestendpointparams"):objectMapper.createObjectNode().toString();
+
+        JsonNode endpointParamsJsonNode;
+        try {
+            endpointParamsJsonNode = objectMapper.readValue(ingestEndpointParams, JsonNode.class);
+        } catch (IOException e) {
+            throw new RuntimeException("Unable to parse ingestendpointparams, cause: " + e.getMessage(), e);
+        }
+
+        boolean doesNotHaveApiPath = !endpointParamsJsonNode.has("apiPath");
+        boolean hasWorkUnitOrEndpointState = endpointParamsJsonNode.has("workUnit") || endpointParamsJsonNode.has("endpointState");
+        if (doesNotHaveApiPath && hasWorkUnitOrEndpointState) {
+            throw new RuntimeException("Cannot set workUnit or endpointState in ingestionendpointparams unless apiPath is defined as well.");
+        }
+
+        final String apiModulePath = (endpointParamsJsonNode.hasNonNull("apiPath") &&
+            endpointParamsJsonNode.get("apiPath").asText().length() > 0) ?
+            endpointParamsJsonNode.get("apiPath").asText() : "/data-hub/5/data-services/ingestion/bulkIngester.api";
+        logger.info("Will write to endpoint defined by: " + apiModulePath);
+        ((ObjectNode)endpointParamsJsonNode).put("apiPath", apiModulePath);
+
+        if (endpointParamsJsonNode.hasNonNull("endpointState") && endpointParamsJsonNode.get("endpointState").asText().length() > 0) {
+            ((ObjectNode)endpointParamsJsonNode).set("endpointState", endpointParamsJsonNode.get("endpointState"));
+        }
+        // TODO : remove the below else block after java-client-api 5.3 release
+        else {
+            ((ObjectNode)endpointParamsJsonNode).put("endpointState", "{}");
+        }
+
+        if (endpointParamsJsonNode.hasNonNull("workUnit") && endpointParamsJsonNode.get("workUnit").asText().length() > 0) {
+
+            JsonNode workUnitNode;
+            try {
+                workUnitNode = objectMapper.readValue(endpointParamsJsonNode.get("workUnit").asText(), JsonNode.class);
+            } catch (JsonProcessingException e) {
+                throw new RuntimeException("Unable to parse workUnit, cause: " + e.getMessage(), e);
+            }
+            ((ObjectNode) workUnitNode).put("uriprefix", params.get("uriprefix"));
+            ((ObjectNode)endpointParamsJsonNode).set("workUnit", workUnitNode);
+        } else {
+            ((ObjectNode)endpointParamsJsonNode).put("workUnit", "{\"uriprefix\":\"" + params.get("uriprefix") + "\"}");
```

Comment:
Do we need to add this even if its null?

Reply:
Yes we need this even when the user did not add URIPrefix because our sjs file looks for this value. Since, the uriPrefix value does not change, we moved it out from endpointState to workUnit

Reply:
Are we going to add null in that case?

Reply:
No we will not be adding null to the uris. Our Glue script will be something like this -

val prefix = ""
val writer = dataframe.write.format("com.marklogic.client.spark.Writer.MarkLogicWriteDataSource")
                    .option("host", host)
                    .option("port", port)
                    .option("user", username)
                    .option("password",password)
                    .option("uriprefix", prefix).....(other options);
By default the value of variable prefix is an empty string. The user can add a custom value.

Reply:
But I don't think Ernie has to specify '.option("uriprefix", prefix)', right? In which case the option will not exist in the options map? And so we'd only want to add it to the workUnit if it's not null (unless the endpoint is careful to not use it if it's null).

Reply:
Yes, that's exactly my point. I don't think that is handled in the endpoint.
Also the endpoint throws an exception if workUnit is empty when no options are set by the user.